### PR TITLE
Metric retrieval recall at precision

### LINKED ISSUE #780: Add `RetrievalRecallAtFixedPrecision`
## 🚀 Feature

Add a new metric calculating (retrieval) recall at K at fixed precision  (typically `K=1` for this case).

### Motivation

From my experience in IR, there are use cases when a user should be provided with some results on top position (e.g. featured snippets in google search) only with some precision threshold. Subsequently, one may be interested in the recall to know which portion of queries will be handled.

### Pitch

Add a new retrieval metric `RetrievalRecallAtFixedPrecision`, which is inspired by `BinnedRecallAtFixedPrecision` and `RetrievalPrecision`.

### PR DESCRIPTION
**Pull request by MrShevan**
## What does this PR do?

Hello, @stancld ! It's a draft PR for issue: Fixes #780

I still think about how to add more tests, testing pattern in this repo is pretty hard, but the general idea of `RetrievalRecallAtFixedPrecision` is implemented.

Keep working 🛠

## Before submitting

- [x] Was this **discussed/approved** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/metrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to **update the docs**?
- [x] Did you write any new **necessary tests**?

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by Borda** [comment]
@MrShevan, how is it going here? :rabbit:

**Comment by MrShevan** [comment]
Hi! Sorry was busy at the end of the week, I keep working on PR. I'm still thinking about my implementation. Was it right to inherit from `Metric` class instead `RetrievalMetric`? I decided to do so to compute recall and precision metrics for all k and return two values: best recall and the value in which it is achieved.

**Comment by stancld** [comment]
Hi @MrShevan, I'm gonna have a look at PR tmrw :]

**Comment by MrShevan** [comment]
Hi @stancld ! I think it was final changes, new metric works and tests are green :] 
Take a look whenever you can.

**Comment by stancld** [comment]
Hi @MrShevan, first of all, thanks a lot for opening this PR! :] I've had a brief look at your implementation and it seems the metric actually does something a bit different than I desired.
My initial thought was to have such a metric where `k` is a priori given parameter, and then recall at precision is calculated. For more context, you may have a ranking algorithm, which sorts paragraphs for each query. Then, for example for `k=1`, you are interested in the recall you are able to deliver given a certain minimum level of precision. For example, Google search sometimes returns featured snippets on the very top of their SERP. Here, you wanna be quite sure the result is correct (i.e. you wanna keep precision >=.95) and then you wanna maximize the recall to target as many users as possible.
The crucial thing thus rather lies in the precision-recall curve instead of the varying `k`.

But let me think for a while if the implementation you propose we can somehow use

**Comment by MrShevan** [comment]
Hi @stancld thanks a lot for your feedback! 
I looked in public sources how the retrieval precision-recall curve is plotted ([https://nlp.stanford.edu/IR-book/pdf/08eval.pdf](url)):
_In a ranked retrieval context, appropriate sets of retrieved documents are naturally given by the top k retrieved documents. For each such set, precision and recall values can be plotted to give a precision-recall curve._

I thought the precision-recall curve is built as points ​​(precision@k, recall@k) for all `k` from `0` to `max_k`, where the `max_k` parameter can be specified. If you specify `max_k=1` in my implementation it seems it will make sense of what you are talking about and you will get the recall value provided that precision >= `min_precision` or an error if this condition is not met for such `k.`

If the parameter `k` is fixed, then only one precision@k value will correspond to `k` and it will not be possible to set conditions for `min_precision.`

Maybe then it is worth to return all the Recall values for each `k` for which the precision >= `min_precision`? Similarly done here in `BinnedRecallAtFixedPrecision` only instead of all values of `k` there are `thresholds`:
[https://github.com/PyTorchLightning/metrics/blob/master/torchmetrics/classification/binned_precision_recall.py#L300](url)
Please correct me if I am wrong in understanding something.

File: torchmetrics/retrieval/recall_precision.py [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+import torch
+from torch import Tensor, tensor
+
+from torchmetrics import Metric
+from torchmetrics.retrieval import RetrievalPrecision, RetrievalRecall
+from torchmetrics.utilities.checks import _check_retrieval_inputs
+from torchmetrics.utilities.data import get_group_indexes
+
+
+class MinPrecisionError(Exception):
```
Comment (Borda) [inline torchmetrics/retrieval/recall_precision.py:26]: lest raise it as ValueError

**Review by Borda — COMMENTED** [review]
lets implement it also as functional and use it in Module

**Comment by MrShevan** [comment]
@Borda Thanks for watching! Got it. I'll implement main logic as functional. 
I think I can reuse the idea of ​​handling ValueError from [here](https://github.com/PyTorchLightning/metrics/blob/master/torchmetrics/classification/binned_precision_recall.py#L35) in case of max() from empty array, then I won't have to write my own implementation of Error :]

File: torchmetrics/retrieval/recall_precision.py [hunk 2]
```diff
@@ -0,0 +1,297 @@
+# Copyright The PyTorch Lightning team.
```
Comment (SkafteNicki) [inline torchmetrics/retrieval/recall_precision.py:1]: please rename this file to `precision_recall_curve.py`

File: torchmetrics/retrieval/recall_precision.py [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+from torchmetrics.utilities.data import get_group_indexes
+
+
+def _retrieval_recall_at_fixed_precision(
+    precision: Tensor,
+    recall: Tensor,
+    top_k: Tensor,
+    min_precision: float,
+) -> Tuple[Tensor, Tensor, Tensor]:
+    try:
```
Comment (SkafteNicki) [inline torchmetrics/retrieval/recall_precision.py:31]: please add docstring here

File: torchmetrics/retrieval/recall_precision.py [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+            - ``'error'``: raise a ``ValueError``
+
+        ignore_index:
+            Ignore predictions where the target is equal to this number.
+        compute_on_step:
+            Forward only calls ``update()`` and returns None if this is set to False.
+
+            .. deprecated:: v0.8
+                Argument has no use anymore and will be removed v0.9.
+
```
Comment (SkafteNicki) [inline torchmetrics/retrieval/recall_precision.py:89]: argument has been removed now
```suggestion
```

File: torchmetrics/retrieval/recall_precision.py [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+    def __init__(
+        self,
+        max_k: Optional[int] = None,
+        adaptive_k: bool = False,
+        empty_target_action: str = "neg",
+        ignore_index: Optional[int] = None,
+        compute_on_step: Optional[bool] = None,
+        **kwargs: Dict[str, Any],
+    ) -> None:
+        super().__init__(compute_on_step=compute_on_step, **kwargs)
```
Comment (SkafteNicki) [inline torchmetrics/retrieval/recall_precision.py:127]:
```suggestion
        **kwargs: Dict[str, Any],
    ) -> None:
        super().__init__(**kwargs)
```

File: torchmetrics/retrieval/recall_precision.py [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+        self.add_state("target", default=[], dist_reduce_fx=None)
+
+        if (max_k is not None) and not (isinstance(max_k, int) and max_k > 0):
+            raise ValueError("`max_k` has to be a positive integer or None")
+
+        if not isinstance(adaptive_k, bool):
+            raise ValueError("`adaptive_k` has to be a boolean")
+
+        self.max_k = max_k
+        self.adaptive_k = adaptive_k
```
Comment (SkafteNicki) [inline torchmetrics/retrieval/recall_precision.py:152]: nitpicking: usually we initialize the states as the last part of `__init__`:
```suggestion
        if (max_k is not None) and not (isinstance(max_k, int) and max_k > 0):
            raise ValueError("`max_k` has to be a positive integer or None")
        self.max_k = max_k
        
        if not isinstance(adaptive_k, bool):
            raise ValueError("`adaptive_k` has to be a boolean")
        self.adaptive_k = adaptive_k
        
        self.add_state("indexes", default=[], dist_reduce_fx=None)
        self.add_state("preds", default=[], dist_reduce_fx=None)
        self.add_state("target", default=[], dist_reduce_fx=None)
```

File: torchmetrics/retrieval/recall_precision.py [hunk 7]
```diff
@@ -0,0 +1,297 @@
...
+
+        self.indexes.append(indexes)
+        self.preds.append(preds)
+        self.target.append(target)
+
+    def compute(self) -> Tuple[Tensor, Tensor, Tensor]:
+        # concat all data
+        indexes = torch.cat(self.indexes, dim=0)
+        preds = torch.cat(self.preds, dim=0)
+        target = torch.cat(self.target, dim=0)
```
Comment (SkafteNicki) [inline torchmetrics/retrieval/recall_precision.py:171]: for consistency, please import and use this function:
https://github.com/PyTorchLightning/metrics/blob/c494f3575f542e7c264555a71fb97e8676f64700/torchmetrics/utilities/data.py#L22-L28

File: torchmetrics/retrieval/recall_precision.py [hunk 8]
```diff
@@ -0,0 +1,297 @@
...
+            - ``'error'``: raise a ``ValueError``
+
+        ignore_index:
+            Ignore predictions where the target is equal to this number.
+        compute_on_step:
+            Forward only calls ``update()`` and returns None if this is set to False.
+
+            .. deprecated:: v0.8
+                Argument has no use anymore and will be removed v0.9.
+
```
Comment (SkafteNicki) [inline torchmetrics/retrieval/recall_precision.py:244]:
```suggestion
```

File: torchmetrics/retrieval/recall_precision.py [hunk 9]
```diff
@@ -0,0 +1,297 @@
...
+        **kwargs: Dict[str, Any],
+    ) -> None:
+        super().__init__(
+            max_k=max_k,
+            adaptive_k=adaptive_k,
+            empty_target_action=empty_target_action,
+            ignore_index=ignore_index,
+            compute_on_step=compute_on_step,
+            **kwargs,
+        )
```
Comment (SkafteNicki) [inline torchmetrics/retrieval/recall_precision.py:287]:
```suggestion
        **kwargs: Dict[str, Any],
    ) -> None:
        super().__init__(
            max_k=max_k,
            adaptive_k=adaptive_k,
            empty_target_action=empty_target_action,
            ignore_index=ignore_index,
            **kwargs,
        )
```

File: torchmetrics/retrieval/recall_precision.py (hunk 2 above)
Comment (MrShevan) [reply torchmetrics/retrieval/recall_precision.py]: renamed 👌

File: torchmetrics/retrieval/recall_precision.py (hunk 3 above)
Comment (MrShevan) [reply torchmetrics/retrieval/recall_precision.py]: added 👌

File: torchmetrics/retrieval/recall_precision.py (hunk 1 above)
Comment (MrShevan) [reply torchmetrics/retrieval/recall_precision.py]: Fixed by func `_retrieval_recall_at_fixed_precision`, which catches all ValueErrors :)

**Comment by Borda** [comment]
> I enjoyed working on this issue, thanks a lot for your involvement :]

great contribution!

**Comment by stancld** [comment]
@MrShevan Thanks a lot for your great contribution! :)
